### PR TITLE
BREAKING CHANGE: remove support for initializing on non-empty directories

### PR DESCRIPTION
**Pull request by shaftoe**
I think the current logic dealing with `.gitignore` is broken, plus, I couldn't understand why we need to update its content if a file with same name is already available in the target folder.

In my opinion we should treat `.gitignore` as every other templated file in order to be idempotent, i.e. if the target file exists than skip creation all together.

### CODE REVIEW

**Comment by shaftoe** [comment]
Ref: https://github.com/probot/create-probot-app/issues/315

**Comment by MaximDevoir** [comment]
Skipping files that are already in the directory makes sense, but the `.gitignore` file should get an exception.

Without amending a pre-existing `.gitignore` file, the user won't know which directories/files they should add unless they were to dig into the source code for CPA.

**Comment by shaftoe** [comment]
I'm not sure I understand what's the feature we want to give to the user. I'd assume that if one is creating a Probot app in a folder which is already being initialized as a git repo and including a `.gitignore` file, he/she already knows how to use git and doesn't need much of a help to ignore whatever should not be committed.

I have the feeling CPA in this case is trying to be smart where there's no need for it. Can you point me to some of the reasons that special logic was implemented?

By the way, as it is now it's not idempotent either: every time CPA is run against the target folder it will append the source `.gitignore` content to the target one.

**Comment by MaximDevoir** [comment]
The feature amends the entries from our `gitignore` onto their existing `.gitignore` file. I think amending `.gitignore` is more of a quality of life feature offered to the user - similar to what's offered by Create React App (Create Next App overwrites .gitignore instead of amending or skipping) - at the cost of a few extra lines of code.

This PR overwrites a pre-existing `.gitignore`. By this nature, the user will have to re-add any entries they had before CPA overwrote `.gitignore`, something the user may not have expected or intended. Additionally, something like `git add * &&  git commit` runs automatically in a later step of CPA. All files that no longer match a gitignore entry have been added to the repo. This requires the user to:

  * remove files from the git repo; and 
  * re-add entries to `.gitignore`.

> Can you point me to some of the reasons that special logic was implemented?

Digging into the git history, I was able to see the logic was added with #73 as a way to properly add `.gitignore` to the CPA-inited directory. That was also where this edge-case logic was added - where I subscribed to CRAs way of handling pre-existing `.gitignore` files (e.g. amending).




**final notes**: I think very few people initiate any of the Create-\*-App (C\*A) on an existing directory. I'd be surprised if anyone has done so with CPA. And every C*A package handles pre-existing files differently. Amending to `.gitignore` is the better solution for handling pre-existing `.gitignore` files - when compared to overwriting or skipping completely - and is a feature we should maintain and improve  (see *side note* below).

____

***side note***: We currently *skip* copying template files if they exist on the target directory, even when they are pertinent files, required for the CPA template to function. This can result in a CPA-inited non-fully-functioning app. This is an issue and can be fixed (i.e: warn the user and abort CPA because files in the directory conflict with files from the CPA template, similar to create-react-app and create-next-app - other solutions exist too) or added to the roadmap.

**Comment by shaftoe** [comment]
> The feature amends the entries from our gitignore onto their existing .gitignore file. I think amending .gitignore is more of a quality of life feature offered to the user - similar to what's offered by Create React App (Create Next App overwrites .gitignore instead of amending or skipping) - at the cost of a few extra lines of code.

I challenge this: personally I'm not happy when I find out that a tool is editing local files without my previous consent. For what I've seen so far we're very careful to keep CPA as less invasive as possible, I still think we should keep that same approach with `.gitignore` too unless there's an expressed consent from the user.

> This PR overwrites a pre-existing .gitignore

Where do you see this happening? From my understanding, it won't, the same logic is applied to every file coming from the templates folder, i.e. overwrite only if expressively set via cli flag, ignore otherwise.

> side note: We currently skip copying template files if they exist on the target directory, even when they are pertinent files, required for the CPA template to function. This can result in a CPA-inited non-fully-functioning app. This is an issue and can be fixed (i.e: warn the user and abort CPA because files in the directory conflict with files from the CPA template, similar to create-react-app and create-next-app - other solutions exist too) or added to the roadmap, if that's something you'd like to look at.
> FYI: From what I can see, Create Next App is able to use the overwrite .gitignore method because CNA doesn't really work on existing directories. CNA will ask the user to remove all files regardless of whether they conflict with the CNA template or not - outside of a few dotfiles and other explicitly allowed files/folders, see here)

In the [latest review](https://github.com/probot/create-probot-app/commit/649063bbeeee0374a91b59eaaa5a976558eef331) of this PR I propose what I believe is a better approach compared to what's in current master (which doesn't meet the expectation you describe above, at least from what I see from some smoke tests):
- if .gitignore not found in target folder, create .gitignore
- else
  - if global overwrite flag is set, overwrite
  - else, warn the user and suggest the amend

**Comment by MaximDevoir** [comment]
Amends we make to .gitignore are logged in the Git repo and can be reverted or changed by the user if needed.

If we do not make changes to .gitignore in an existing Git repo, what should we do about the files that are automatically committed to .gitignore during app initialization (e.g. secrets/`.env` files, `node_modules` would be committed if not previously *gitignored*, and other files now committed)? Logging what to ignore in the console when it has already been committed seems like a regression in QoL.

Since this topic has come up, my [side note from above](https://github.com/probot/create-probot-app/pull/339#issuecomment-764198761) describes how CPA *really* lacks support for initializing onto existing directories. I don't see anyone using CPA on existing directories. If they were, then we probably would've gotten a few issues describing how *broken* the tool is when initializing onto existing directories or Git repos.

I would be okay with removing support for initializing onto existing directories - removing the extra gitignore logic - for now. Then, if someone wanted to take the problem of handling initializing onto pre-existing directories/repos, properly handling application-critical conflicting files, and dealing with pre-existing `.gitignore`, then we could revisit the issue.

**Comment by shaftoe** [comment]
> Amends we make to .gitignore are logged in the Git repo and can be reverted or changed by the user if needed.

What you say above assumes a target clean working tree at the time CPA is running. Just to point out that there might be many other edge cases to think about when manipulating already created files.

> If we do not make changes to .gitignore in an existing Git repo, what should we do about the files that are automatically committed to .gitignore during app initialization (e.g. secrets/`.env` files, `node_modules` would be committed if not previously _gitignored_, and other files now committed)? Logging what to ignore in the console when it has already been committed seems like a regression in QoL.

I think we should skip completely `git commit` (and generally, any automatic git handling) in case the target folder is already present. Even better, we should require a non-existing target folder and (maybe) keep the overwrite logic to make the user acknowledge he's going out of the recommended workflow.

I discourage the implementation of any amending logic of any file, it should be considered out of scope of CPA, it's complex logic with lots of edge cases to deal with and, most importantly, arguably it doesn't provide any real benefit to most of CPA users anyway.

**Comment by MaximDevoir** [comment]
> What you say above assumes a target clean working tree at the time CPA is running. Just to point out that there might be many other edge cases to think about when manipulating already created files.

You're right. This PR has brought to my attention how unsuitable CPA is for handling pre-existing directories.

> Even better, we should require a non-existing target folder

I agree. I am okay with getting rid of support for targeting existing directories.

@shaftoe Do you want to supersede this PR with one that removes support for targeting existing directories?

**Comment by shaftoe** [comment]
> You're right. This PR has brought to my attention how unsuitable CPA is for handling pre-existing directories.

@MaximDevoir  There's only one doubt left for me: should we consider that a breaking change? i.e. bump the major version too?

> I agree. I am okay with getting rid of support for targeting existing directories.
> 
> @shaftoe Do you want to supersede this PR with one that removes support for targeting existing directories?

Yeah, please disregard the current implementation, I'll come up with one which does what we just agreed and supersede the current patch.

**Comment by MaximDevoir** [comment]
> should we consider that a breaking change?

We can major bump :+1:

**Comment by MaximDevoir** [comment]
Awesome, I'm on mobile and I'll be able to review this when I'm available later today.

File: src/create-probot-app.ts [hunk 1]
```diff
@@ -2,13 +2,14 @@
...
 import { installAndBuild } from "./helpers/run-npm";
-import { makeScaffolding } from "./helpers/filesystem";
+import { ensureValidDestination, makeScaffolding } from "./helpers/filesystem";
 import { printSuccess, red } from "./helpers/write-help";
 
 async function main(): Promise<void> {
   const config = await askUser(runCliManager());
 
-  makeScaffolding(config)
+  ensureValidDestination(config.destination, config.overwrite)
```
Comment (MaximDevoir) [inline src/create-probot-app.ts:11]: The inquirer prompts will be asked, and then an error will be thrown at the end, if the destination was invalid. We should validate the destination as early as possible.
Could be done by moving the `ensureValidDestination` check to `user-interation.ts`. as shown below.


```diff
diff --git a/src/helpers/user-interaction.ts b/src/helpers/user-interaction.ts
index cc135e6..5aca32b 100644
--- a/src/helpers/user-interaction.ts
+++ b/src/helpers/user-interaction.ts
@@ -218,6 +218,7 @@ export function runCliManager(): CliConfig {
   const options = program.parse(process.argv).opts();
 
   if (!destination) printHelpAndFail();
+  if(!ensureValidDestination()) printDesinationHelpOrWhateverAndFail();
   if (options.showTemplates) {
     getTemplates().forEach((template) => console.log(template));
     process.exit();
```

File: src/helpers/filesystem.ts [hunk 2]
```diff
@@ -6,6 +6,50 @@ import { yellow, green } from "./write-help";
...
+      console.warn(
+        yellow(`Explicit OVERWRITE option selected:
+Some files under "${fs.realpathSync(destination)}" might be overwritten!
+`)
+      );
+      return;
+    }
+  }
+
+  // either not a folder or a not empty one
```
Comment (MaximDevoir) [inline src/helpers/filesystem.ts:49]: nit: Add a subject for clarification.

```suggestion
  // The `destination` is either not a folder or a not empty one.
```

File: src/helpers/filesystem.ts [hunk 3]
```diff
@@ -6,6 +6,50 @@ import { yellow, green } from "./write-help";
...
+
+    try {
+      // edge case: destination might be a broken link
+      fs.realpathSync(destination);
+    } catch (error) {
+      if (error.code === "ENOENT") throw invalidDestinationError;
+    }
+  } catch (error) {
+    if (error.code === "ENOENT") return;
+    // not-existing path
```
Comment (MaximDevoir) [inline src/helpers/filesystem.ts:33]: nit: Add subject to the sentence. What isn't an existing path, the destination?

**Review by MaximDevoir — APPROVED** [review]

**Review by gr2m — APPROVED** [review]
Great work both of you 👍🏼

**Pull request merged**